app/controllers/main/routes.py

```diff
@@ -37,38 +40,72 @@ def events(selectedTerm=None):
         currentTime = currentTime,
         user = g.current_user)
 
-@main_bp.route('/profile/<username>', methods = ['GET'])
-def profilePage(username):
-    if not g.current_user.isCeltsAdmin and not g.current_user.isCeltsStudentStaff and g.current_user.username != username:
-        return "Access Denied", 403
+@main_bp.route('/profile/<username>', methods=['GET'])
+def viewVolunteersProfile(username):
+    """
+    This function displays the information of a volunteer to the user
+    """
+    if (g.current_user.username == username) or g.current_user.isAdmin:
+         upcomingEvents = getUpcomingEventsForUser(username)
+         programs = Program.select()
+         interests = Interest.select().where(Interest.user == username)
+         programBan = ProgramBan.select().where(ProgramBan.user == username)
+         interests_ids = [interest.program for interest in interests]
+         eventParticipant = EventParticipant.select().where(EventParticipant.user == username)
+         currentTerm = Term.select().where(Term.isCurrentTerm == 1)
+         trainingEvents = (ProgramEvent.select().join(Event).where((Event.term == currentTerm) & (Event.isTraining == 1)))
+         trainingChecklist = {}
+         for program in programs:
+             trainingChecklist[program.id] = trainedParticipants(program.id)
```

Comment:
This is a lot of work (a query for each program). It seems like it's coming at the problem a little backwards. You're getting a list of everyone who has completed the training for each program, rather than just looking to see if the volunteer has completed the trainings in each program.

Reply:
Deleted this part as it is being done later in the file in the eligibilityTable, will try to fix the query there, instead.